contracts/MigrationContractStaging.cdc

```diff
@@ -0,0 +1,362 @@
+            ?? panic("Problem destroying update Capsule")
+        emit StagingStatusUpdated(
+            capsuleUUID: capsuleUUID,
+            address: address,
+            codeHash: [],
+            contract: name,
+            action: "unstage"
+        )
+    }
+
+    /* --- Public Getters --- */
+
+    /// Returns true if the contract is currently staged.
+    ///
+    access(all) view fun isStaged(address: Address, name: String): Bool {
+        return self.stagedContracts[address]?.contains(name) ?? false
+    }
+
+    /// Returns the names of all staged contracts for the given address.
+    ///
+    access(all) view fun getStagedContractNames(forAddress: Address): [String] {
+        return self.stagedContracts[forAddress] ?? []
+    }
+
+    /// Returns the staged contract Cadence code for the given address and name.
+    ///
+    access(all) fun getStagedContractCode(address: Address, name: String): String? {
+        let capsulePath: StoragePath = self.deriveCapsuleStoragePath(contractAddress: address, contractName: name)
+        if let capsule = self.account.borrow<&Capsule>(from: capsulePath) {
+            return capsule.getContractUpdate().codeAsCadence()
+        } else {
+            return nil
+        }
+    }
+
+    /// Returns an array of all staged contract host addresses.
+    ///
+    access(all) view fun getAllStagedContractHosts(): [Address] {
+        return self.stagedContracts.keys
+    }
+
+    /// Returns a dictionary of all staged contract code for the given address.
+    ///
+    access(all) view fun getAllStagedContractCode(forAddress: Address): {String: String} {
+        if self.stagedContracts[forAddress] == nil {
+            return {}
+        }
+        let capsulePaths: [StoragePath] = []
+        let stagedCode: {String: String} = {}
+        let contractNames: [String] = self.stagedContracts[forAddress]!
+        for name in contractNames {
+            capsulePaths.append(self.deriveCapsuleStoragePath(contractAddress: forAddress, contractName: name))
+        }
+        for path in capsulePaths {
+            if let capsule = self.account.borrow<&Capsule>(from: path) {
+                let update: ContractUpdate = capsule.getContractUpdate()
+                stagedCode[update.name] = update.codeAsCadence()
+            }
+        }
+        return stagedCode
+    }
+
+    /// Returns a StoragePath to store the Capsule of the form:
+    ///     /storage/self.capsulePathPrefix_ADDRESS_NAME
+    access(all) view fun deriveCapsuleStoragePath(contractAddress: Address, contractName: String): StoragePath {
+        return StoragePath(
+                identifier: self.capsulePathPrefix
+                    .concat(self.delimiter)
+                    .concat(contractAddress.toString())
+                    .concat(self.delimiter)
+                    .concat(contractName)
+            ) ?? panic("Could not derive Capsule StoragePath for given address")
```

Comment:
Introduce a variable to avoid complex nesting: 
```suggestion
        let identifier = self.capsulePathPrefix
            .concat(self.delimiter)
            .concat(contractAddress.toString())
            .concat(self.delimiter)
            .concat(contractName)
        return StoragePath(identifier: identifier)
            ?? panic("Could not derive Capsule StoragePath for given address")
```